usecase/src/main/java/org/mobilitydata/gtfsvalidator/usecase/ValidateNoOverlappingStopTimeInTripBlock.java

```diff
@@ -0,0 +1,264 @@
+ * Use case to validate that for a given `trip_id` there is no overlapping {@code StopTime}.
+ */
+public class ValidateNoOverlappingStopTimeInTripBlock {
+    private final GtfsDataRepository dataRepo;
+    private final ValidationResultRepository resultRepo;
+    private final Logger logger;
+    private final TimeUtils timeUtils;
+
+    public ValidateNoOverlappingStopTimeInTripBlock(final GtfsDataRepository dataRepo,
+                                                    final ValidationResultRepository resultRepo,
+                                                    final Logger logger,
+                                                    final TimeUtils timeUtils) {
+        this.dataRepo = dataRepo;
+        this.resultRepo = resultRepo;
+        this.logger = logger;
+        this.timeUtils = timeUtils;
+    }
+
+    /**
+     * Use case execution method: checks for each trip of the same block that no stop times overlap:
+     * - trips from same block with same service id that overlap should generate a notice
+     * - trips from same block with same service_id that do not overlap should not generate notice
+     * - trips from same block with with different service_id that overlap should generate notice whether calendar or
+     * calendar_Dates is provided
+     * - trips from same block with with different service_id that do not overlap should not generate notice whether
+     * calendar or calendar_dates is provided
+     */
+    public void execute() {
+        logger.info("validating rule 'E054 - Trips from same block overlap'");
+
+        final Map<String, List<Trip>> tripPerBlockId = dataRepo.getAllTripByBlockId();
+        final boolean areCalendarProvided = dataRepo.getCalendarAll().size() != 0;
+
+        tripPerBlockId.forEach((blockId, tripCollectionFromSameBlockId) -> {
+            final List<String> visitedTripIdCollection = new ArrayList<>();
+            tripCollectionFromSameBlockId.forEach(currentTrip -> {
+
+                final String currentTripId = currentTrip.getTripId();
+                final SortedMap<Integer, StopTime> stopTimeCollectionForCurrentTrip =
+                        dataRepo.getStopTimeByTripId(currentTrip.getTripId());
+
+                final StopTime currentTripFirstStopTimeInSequence =
+                        stopTimeCollectionForCurrentTrip.get(stopTimeCollectionForCurrentTrip.firstKey());
+                final StopTime currentTripLastStopTimeInSequence =
+                        stopTimeCollectionForCurrentTrip.get(stopTimeCollectionForCurrentTrip.lastKey());
+                final Integer currentTripFirstTime = currentTripFirstStopTimeInSequence.getArrivalTime();
+                final Integer currentTripLastTime = currentTripLastStopTimeInSequence.getDepartureTime();
+
+                visitedTripIdCollection.add(currentTripId);
+
+                if (currentTripFirstTime != null && currentTripLastTime != null) {
+                    if (currentTripFirstTime <= currentTripLastTime) {
```

Comment:
The trip first and last time can't be equal, right? Otherwise the trip is `0` seconds long?

Also, if this is another sanity check, I'd suggest refactoring to `areTripTimesValid()` method to better explain what's happening.

Reply:
https://github.com/MobilityData/gtfs-validator/pull/346#discussion_r481122879